### PR TITLE
[android] reduce Java interop for `ContentViewGroup` and `WrapperView`

### LINKED ISSUE #9485: [Android] Slow scrolling, laggy performance
### Description

It is a list using CollectionView with elements using Border. It's a list of only 5 items, but there must be something wrong somewhere because it lags considerably.

surfer-droid-scroll

### Steps to Reproduce

1. Download or sync https://github.com/jsuarezruiz/netmaui-surfing-app-challenge
2. Launch the App on Android and scroll the main page.

### Version with bug

6.0.486 (current)

### Last version that worked well

Unknown/Other

### Affected platforms

Android

### Affected platform versions

Android 12 and 13

### Did you find any workaround?

_No response_

### Relevant log output

_No response_

### PR DESCRIPTION
**Pull request by jonathanpeppers**
Fixes #9485
Context: https://github.com/jsuarezruiz/netmaui-surfing-app-challenge
Context: https://developer.android.com/reference/androidx/core/app/FrameMetricsAggregator#getMetrics()
Context: https://medium.com/grindr-engineering/android-metrics-framerate-for-each-fragments-d311dbe909c1

Reviewing the "Surfing App" sample, I noticed that a lot of time is spent doing Java interop when scrolling:

    1.76s (35%) Microsoft.Maui!Microsoft.Maui.Platform.WrapperView.DispatchDraw(Android.Graphics.Canvas)
    1.76s (35%) Microsoft.Maui!Microsoft.Maui.Platform.ContentViewGroup.DispatchDraw(Android.Graphics.Canvas)

This methods were deeply nested doing interop from Java -> C# -> Java many levels deep. My thought is we can probably move some code from C# to Java, in order to do a lot less interop while scrolling. I created new `PlatformContentViewGroup` and `PlatformWrapperView` Java types for this.

So to clarify, we go from:

* `ContentViewGroup`
  * `DispatchDraw(Canvas)` Java -> C#
  * `DispatchDraw(Canvas)` does a lookup for living C# instances of `Canvas`
* `ContentViewGroup`
  * `DispatchDraw(Canvas)` Java -> C# (only when `Clip` is set)
  * `GetClipPath(int, int)` has the width/height and no `Canvas`
* `WrapperView`
  * `ctor(Context)` makes several C# -> Java calls
  * `DispatchDraw(Canvas)` Java -> C#
  * `DrawShadow(Canvas)` makes several C# -> Java calls for width & height
  * `OnMeasure()` Java -> C#
* `WrapperView`
  * `ctor(Context)` does all the work in Java now
  * `DispatchDraw(Canvas)` Java -> C# (only when `Clip` or `Shadow` are set)
  * `GetClipPath(int, int)` has the width/height and no `Canvas`
  * `DrawShadow(Canvas, int, int)` passes in the width/height from Java
  * `OnMeasure()` stays completely in Java

So then, how do we measure if it is better/worse? I used the `FrameMetricsAggregator` class we can add to any app's `MainActivity`:

```csharp
const int Duration = 1000;
FrameMetricsAggregator aggregator;
Handler handler;

protected override void OnCreate(Bundle savedInstanceState)
{
    base.OnCreate(savedInstanceState);

    handler = new Handler(Looper.MainLooper);
    aggregator = new FrameMetricsAggregator(FrameMetricsAggregator.TotalDuration);
    aggregator.Add(this);

    handler.PostDelayed(OnFrame, Duration);
}

void OnFrame()
{
    var metrics = aggregator.GetMetrics()[FrameMetricsAggregator.TotalIndex];
    int size = metrics.Size();
    double sum = 0, count = 0, slow = 0;
    for (int i = 0; i < size; i++)
    {
        int value = metrics.Get(i);
        if (value != 0)
        {
            count += value;
            sum += i * value;
            if (i > 16)
                slow += value;
            Console.WriteLine($"Frame(s) that took ~{i}ms, count: {value}");
        }
    }
    if (sum > 0)
    {
        Console.WriteLine($"Average frame time: {sum / count:0.00}ms");
        Console.WriteLine($"No. of slow frames: {slow}");
        Console.WriteLine("-----");
    }
    handler.PostDelayed(OnFrame, Duration);
}
```

I added many more rows to the sample, and scrolled to the bottom on a `Release` build of the app on a Pixel 5:

    Before:
    Frame(s) that took ~4ms, count: 1
    Frame(s) that took ~5ms, count: 6
    Frame(s) that took ~6ms, count: 10
    Frame(s) that took ~7ms, count: 12
    Frame(s) that took ~8ms, count: 10
    Frame(s) that took ~9ms, count: 6
    Frame(s) that took ~10ms, count: 1
    Frame(s) that took ~11ms, count: 2
    Frame(s) that took ~12ms, count: 4
    Frame(s) that took ~13ms, count: 2
    Frame(s) that took ~15ms, count: 1
    Frame(s) that took ~16ms, count: 1
    Frame(s) that took ~18ms, count: 2
    Frame(s) that took ~19ms, count: 1
    Frame(s) that took ~20ms, count: 5
    Frame(s) that took ~21ms, count: 2
    Frame(s) that took ~22ms, count: 1
    Frame(s) that took ~25ms, count: 1
    Frame(s) that took ~32ms, count: 1
    Frame(s) that took ~34ms, count: 1
    Frame(s) that took ~60ms, count: 1
    Frame(s) that took ~62ms, count: 1
    Frame(s) that took ~63ms, count: 1
    Frame(s) that took ~64ms, count: 2
    Frame(s) that took ~66ms, count: 1
    Frame(s) that took ~67ms, count: 1
    Frame(s) that took ~68ms, count: 1
    Frame(s) that took ~69ms, count: 2
    Frame(s) that took ~70ms, count: 2
    Frame(s) that took ~71ms, count: 2
    Frame(s) that took ~72ms, count: 1
    Frame(s) that took ~73ms, count: 2
    Frame(s) that took ~74ms, count: 2
    Frame(s) that took ~75ms, count: 1
    Frame(s) that took ~76ms, count: 1
    Frame(s) that took ~77ms, count: 2
    Frame(s) that took ~78ms, count: 3
    Frame(s) that took ~79ms, count: 1
    Frame(s) that took ~80ms, count: 1
    Frame(s) that took ~81ms, count: 1
    Average frame time: 28.67ms
    No. of slow frames: 43

    After:
    Frame(s) that took ~5ms, count: 3
    Frame(s) that took ~6ms, count: 5
    Frame(s) that took ~7ms, count: 7
    Frame(s) that took ~8ms, count: 7
    Frame(s) that took ~9ms, count: 4
    Frame(s) that took ~10ms, count: 2
    Frame(s) that took ~11ms, count: 6
    Frame(s) that took ~12ms, count: 2
    Frame(s) that took ~13ms, count: 3
    Frame(s) that took ~14ms, count: 4
    Frame(s) that took ~15ms, count: 1
    Frame(s) that took ~16ms, count: 1
    Frame(s) that took ~17ms, count: 1
    Frame(s) that took ~18ms, count: 2
    Frame(s) that took ~19ms, count: 1
    Frame(s) that took ~20ms, count: 3
    Frame(s) that took ~21ms, count: 2
    Frame(s) that took ~22ms, count: 2
    Frame(s) that took ~27ms, count: 2
    Frame(s) that took ~29ms, count: 2
    Frame(s) that took ~32ms, count: 1
    Frame(s) that took ~34ms, count: 1
    Frame(s) that took ~35ms, count: 1
    Frame(s) that took ~64ms, count: 1
    Frame(s) that took ~67ms, count: 1
    Frame(s) that took ~68ms, count: 2
    Frame(s) that took ~69ms, count: 1
    Frame(s) that took ~72ms, count: 3
    Frame(s) that took ~74ms, count: 3
    Average frame time: 21.99ms
    No. of slow frames: 29

A slow frame is defined as one that takes longer than 16ms (if you drop below 60 fps). On a Pixel 5 device, this change went from 43 to 29 slow frames.

I think the app looks the exact same:

image

Note that the play button icon is not visible in the screenshot because of a separate issue, see: https://github.com/dotnet/maui/issues/14239

I think there is enough improvement here to close #9485, but we can always do more. This at least makes things a lot better for all `View`'s on Android as well as other types of views like `Border` and `Frame`.

### CODE REVIEW

**Review by jsuarezruiz — APPROVED** [review]
With the changes, will use directly java except if use Shadows and Clipping. Maybe we can continue improving the numbers with changes in that cases, but this is a nice progress. LGTM!

File: src/Core/AndroidNative/maui/src/main/java/com/microsoft/maui/PlatformWrapperView.java [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+
+    @Override
+    protected void dispatchDraw(Canvas canvas) {
+        // Only call into C# if there is a Shadow
+        if (hasShadow) {
+            int viewWidth = viewBounds.width();
+            int viewHeight = viewBounds.height();
+            if (getChildCount() > 0)
+            {
+                View child = getChildAt(0);
```
Comment (mandel-macaque) [inline src/Core/AndroidNative/maui/src/main/java/com/microsoft/maui/PlatformWrapperView.java:51]: Is it guaranteed here that we do have at lest a child? Could it be empty and result in an NRE?

**Review by mandel-macaque — APPROVED** [review]
LGTM but I had a question about the getChildAt method which I know can return null but no tif that is possible in the code at runtime.

**Review by hartez — APPROVED** [review]

**Comment by jonathanpeppers** [comment]
> @jonathanpeppers can we solve the .aar conflict?

Should be rebased now, and I ran `dotnet cake --target=android-aar --configuration=Release`.

**Review by rmarinho — APPROVED** [review]

**Pull request merged**